uniffi_bindgen/src/bindings/gecko/gen_gecko.rs

```diff
@@ -0,0 +1,355 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::path::Path;
+
+use anyhow::Result;
+use askama::Template;
+use heck::{CamelCase, MixedCase};
+
+use crate::interface::*;
+
+// Some config options for the caller to customize the generated Gecko bindings.
+// Note that this can only be used to control details *that do not affect the
+// underlying component*, since the details of the underlying component are
+// entirely determined by the `ComponentInterface`.
+pub struct Config {
+    // ...
+}
+
+impl Config {
+    pub fn from(_ci: &ComponentInterface) -> Self {
+        Config {
+            // ...
+        }
+    }
+}
+
+/// Indicates whether a WebIDL type is reflected as an out parameter or return
+/// value in C++. This is used by the namespace and interface templates to
+/// generate the correct argument lists for the binding.
+pub enum ReturnPosition<'a> {
+    OutParam(&'a Type),
+    Return(&'a Type),
+}
+
+impl<'a> ReturnPosition<'a> {
+    /// Given a type, returns a tag indicating whether it's returned by value or
+    /// via an out parameter.
+    pub fn for_type(type_: &'a Type) -> ReturnPosition<'a> {
+        match type_ {
+            Type::String => ReturnPosition::OutParam(type_),
+            Type::Optional(_) => ReturnPosition::OutParam(type_),
+            Type::Record(_) => ReturnPosition::OutParam(type_),
+            Type::Map(_) => ReturnPosition::OutParam(type_),
+            Type::Sequence(_) => ReturnPosition::OutParam(type_),
+            _ => ReturnPosition::Return(type_),
+        }
+    }
+}
+
+/// Returns a dummy empty value for the given type. This is used to
+/// implement the `cpp::bail` macro to return early if a WebIDL function or
+/// method throws an exception.
+pub fn default_return_value_cpp(type_: &Type) -> Option<String> {
```

Comment:
It’s unfortunate that this has to exist. Basically, if you want to throw an error from inside a WebIDL method with a signature like `int32_t Foo(ErrorResult& aRv)`, you can call `aRv.Throw(...)` in the method, but you still need to return _some_ value—even though it’ll get ignored by the bindings code. Methods that return results via out params don’t need to do this; they can just do `return;` and not set the param.